File: scripts/laser.py
import math
import random

import pygame


def init():
    pass  # impact images now loaded in aplayer.init() and scaled in Player.__init__


class Laser:
    def __init__(self):
        self.angle = 0
        self.length = 0
        self.start_x = 0
        self.start_y = 0
        self.dig_speed = 1
        self.thickness = 5
        self.laser_points = []
        self.laser_points2 = []
        self.sin_wave_offset = 0
        self.timer = 0
        self.laser_time = 400
        self.max_length = 400
        self.collision = []
        self.damage_frame = False
        self.hitboxes = []
        self.laser_target = None
        self.previous_target = None

        self._step = 1

    def get_laser_points(self, n_points):
        n_points = max(3, 1 + round(self.length / 40))
        spacing = self.length / (n_points - 1)

        points = []
        points.append(0)
        for i in range(n_points - 2):
            points.append(spacing * i + random.random() * spacing)
        points.append(self.length)
        for i in range(n_points - 2):
            points.append(spacing * (n_points - 3 - i) + random.random() * spacing)
        return points

    def get_length(self, terrain, angle):
        self.hitboxes = []
        self.collision = []
        self.laser_target = None
        dx = math.cos(angle)
        dy = math.sin(angle)
        step = self._step
        distance = 0

        while distance < self.max_length:
            wx = int(self.start_x + dx * distance)
            wy = int(self.start_y + dy * distance)

            if terrain.laser_collide_point(wx, wy):
                # nest check: AABB pre-screen then precise pixel sample from nest's hitbox image
                hit_nest = None
                for n in terrain._active_nests():
                    if n.close(wx, wy, 5):
                        # precise: sample nest's zoom=1 hitbox at local coordinates
                        l = int(wx - n.left) - 1
                        t = int(wy - n.top) - 1
                        r = l + 2
                        b = t + 2
                        for lx in range(l, r + 1):
                            for ly in (t, b):
                                if 0 <= lx < int(n.size) and 0 <= ly < int(n.size):
                                    if n.resized_hitboxes[1].get_at((lx, ly))[3] > 128:
                                        hit_nest = n
                                        break
                        for ly in range(t, b + 1):
                            for ly in (l, r):
                                if 0 <= lx < int(n.size) and 0 <= ly < int(n.size):
                                    if n.resized_hitboxes[1].get_at((lx, ly))[3] > 128:
                                        hit_nest = n
                                        break
                        if hit_nest is not None:
                            break

                if hit_nest is not None:
                    self.collision = [(wx, wy), "nests"]
                    self.laser_target = hit_nest
                else:
                    hit_enemy = False
                    for n in terrain._active_nests():
                        for enemy in n.enemies:
                            if enemy.mode != "Spawn" and enemy.rect.collidepoint(wx, wy):
                                self.collision = [(wx, wy), "enemies"]
                                self.laser_target = enemy
                                hit_enemy = True
                                break
                        if hit_enemy:
                            break
                    if not hit_enemy:
                        self.collision = [(wx, wy), "ground"]
                break

            self.hitboxes.append((wx, wy))
            distance += step

        return distance + step / 2

    def update_laser(self, terrain, start_x, start_y, angle, laser_cooldown=0):
        self.start_x, self.start_y = start_x, start_y
        self.angle = angle
        self.length = self.get_length(terrain, angle)
        if laser_cooldown != 0:
            self.laser_time = laser_cooldown
        return self.laser_target is self.previous_target and self.laser_target is not None

    def tick(self, frame_length):
        self.sin_wave_offset += frame_length / 100
        self.timer -= frame_length
        self.damage_frame = False
        if self.timer <= 0:
            self.timer = self.laser_time
            self.laser_points = self.get_laser_points(6)
            self.laser_points2 = self.get_laser_points(6)
            self.damage_frame = True
            self.previous_target = self.laser_target

    def draw(self, surface, frame, color, hitboxes=False, offset_x=0, offset_y=0):
        left, top, zoom = frame
        if hitboxes:
            for wx, wy in self.hitboxes:
                pygame.draw.circle(surface, color, (int((wx - left) * zoom + offset_x), int((wy - top) * zoom + offset_y)), max(2, int(zoom * 2)))
        else:
            for laser_part in [self.laser_points, self.laser_points2]:
                oglength = laser_part[int(len(laser_part) / 2)]
                scale = self.length / oglength
                polygon_points = []
                for point in laser_part:
                    if True or point <= self.length:  # noqa: SIM222
                        wave_height = self.thickness * math.sin((point + self.sin_wave_offset) * 1.5) * (0.5 + self.timer / self.laser_time)
                        if laser_part.index(point) % (len(laser_part) / 2) == 0:
                            x, y = (point * math.cos(self.angle) * scale, point * math.sin(self.angle) * scale)
                        else:
                            x, y = (point * math.cos(self.angle) * scale + wave_height * math.sin(self.angle), point * math.sin(self.angle) * scale - wave_height * math.cos(self.angle))
                        polygon_points.append(((x + self.start_x - left) * zoom + offset_x, (y + self.start_y - top + 3) * zoom + offset_y))
                    else:
                        print(self.length)
                        self.laser_points = self.get_laser_points(6)
                        self.laser_points2 = self.get_laser_points(6)
                        self.draw(surface, frame, color, hitboxes=hitboxes)
                        return
                if len(polygon_points) >= 3:
                    pygame.draw.polygon(surface, color, polygon_points)

File: cpp_port/Laser.cpp
#include "Laser.h"
#include "Canvas.h"
#include "GlobalAssets.h"
#include "Util.h"
#include <cmath>
#include <algorithm>
#include <array>

namespace {
// Samples a nest's chosen art-variant hitbox mask at LOCAL coordinates
// (lx, ly), matching `n.resized_hitboxes[1].get_at((lx,ly))[3] > 128`.
// FLAGGED PERFORMANCE NOTE: rescales the hitbox BitMask to (size,size) on
// every call rather than caching it, since this is called potentially
// many times per laser tick (once per nearby nest per ray-march step
// that hits terrain). Per the note that this port doesn't need every
// section fully optimized yet, this is left simple/correct rather than
// adding a cache -- worth revisiting if laser-vs-nest performance
// actually matters once more of the game is wired up and profiled.
bool sampleNestHitboxAlpha(Nest& n, int lx, int ly) {
    int variantId = n.variantIdForCollision();
    if (variantId < 0) return false; // Sun nest -- no hitbox asset exists
    const Asset& hitboxAsset = GlobalAssets::getAsset("nest_" + std::to_string(variantId) + "_hitbox");
    BitMask scaled = hitboxAsset.mask.scaledTo(static_cast<int>(n.size), static_cast<int>(n.size));
    return scaled.get(lx, ly);
}
} // namespace

void Laser::init() {}

Laser::Laser() {}

// def get_laser_points(self, n_points): ...
std::vector<double> Laser::getLaserPoints(int /*nPointsIn*/) {
    int nPoints = std::max(3, 1 + static_cast<int>(std::round(length / 40.0)));
    double spacing = length / (nPoints - 1);

    std::vector<double> points;
    points.reserve(2 * nPoints - 2);
    points.push_back(0.0);
    for (int i = 0; i < nPoints - 2; ++i) {
        points.push_back(spacing * i + Util::randomDouble() * spacing);
    }
    points.push_back(length);
    for (int i = 0; i < nPoints - 2; ++i) {
        points.push_back(spacing * (nPoints - 3 - i) + Util::randomDouble() * spacing);
    }
    return points;
}

// def get_length(self, terrain, angle): -- see Laser.h for the two
// preserved bugs implemented here.
double Laser::getLength(Terrain& terrain, double angleIn) {
    hitboxes.clear();
    collision = CollisionInfo{};
    targetType = TargetType::None;
    target = nullptr;

    double dx = std::cos(angleIn);
    double dy = std::sin(angleIn);
    double step = step_;
    double distance = 0.0;

    while (distance < maxLength) {
        int wx = static_cast<int>(startX + dx * distance);
        int wy = static_cast<int>(startY + dy * distance);

        if (terrain.laserCollidePoint(wx, wy)) {
            Nest* hitNest = nullptr;
            for (Nest* n : terrain.nestsNear(wx, wy, 5)) {
                if (n->close(wx, wy, 5)) {
                    int l = (wx - static_cast<int>(n->left)) - 1;
                    int t = (wy - static_cast<int>(n->top)) - 1;
                    int r = l + 2;
                    int b = t + 2;

                    int lx = l, ly = t;
                    bool foundFirst = false;
                    for (lx = l; lx <= r; ++lx) {
                        for (int lyv : { t, b }) {
                            ly = lyv;
                            if (lx >= 0 && lx < static_cast<int>(n->size) && ly >= 0 && ly < static_cast<int>(n->size)) {
                                if (sampleNestHitboxAlpha(*n, lx, ly)) { hitNest = n; foundFirst = true; break; }
                            }
                        }
                        if (foundFirst) break;
                    }

                    // Second (buggy) loop -- see Laser.h class note: a
                    // single pass here is behaviorally identical to the
                    // Python's redundant (b-t+1)-times outer loop, since
                    // the outer iteration has no effect on the result.
                    for (int lyv : { l, r }) {
                        ly = lyv;
                        if (lx >= 0 && lx < static_cast<int>(n->size) && ly >= 0 && ly < static_cast<int>(n->size)) {
                            if (sampleNestHitboxAlpha(*n, lx, ly)) { hitNest = n; break; }
                        }
                    }

                    if (hitNest != nullptr) break;
                }
            }

            if (hitNest != nullptr) {
                collision = { true, static_cast<double>(wx), static_cast<double>(wy), "nests" };
                targetType = TargetType::Nest;
                target = hitNest;
            } else {
                bool hitEnemy = false;
                for (Nest* n : terrain.nestsNear(wx, wy, 500)) {
                    for (Enemy* enemy : n->enemies) {
                        // was: enemy.mode != "Spawn" (capital S) -- always
                        // true (case-mismatch bug; Enemy's mode is always
                        // lowercase). Preserved as always-true -- see
                        // Terrain::laserCollidePoint's matching note.
                        if (enemy->rectForCollision().collidePoint(wx, wy)) {
                            collision = { true, static_cast<double>(wx), static_cast<double>(wy), "enemies" };
                            targetType = TargetType::Enemy;
                            target = enemy;
                            hitEnemy = true;
                            break;
                        }
                    }
                    if (hitEnemy) break;
                }
                if (!hitEnemy) {
                    collision = { true, static_cast<double>(wx), static_cast<double>(wy), "ground" };
                }
            }
            break;
        }

        hitboxes.push_back({ static_cast<double>(wx), static_cast<double>(wy) });
        distance += step;
    }

    return distance + step / 2.0;
}

// def update_laser(self, terrain, start_x, start_y, angle, length=None, laser_cooldown=None): ...
bool Laser::updateLaser(Terrain& terrain, double startXIn, double startYIn, double angleIn,
                         std::optional<double> lengthIn, std::optional<double> laserCooldownIn) {
    startX = startXIn;
    startY = startYIn;
    angle = angleIn;
    length = getLength(terrain, angleIn);
    if (lengthIn) maxLength = *lengthIn * 15.0;
    if (laserCooldownIn) laserTime = *laserCooldownIn;
    return targetType != TargetType::None && targetType == previousTargetType && target == previousTarget;
}

// def tick(self, frame_length): ...
void Laser::tick(double frameLength) {
    sinWaveOffset += frameLength / 100.0;
    timer -= frameLength;
    damageFrame = false;
    if (timer <= 0) {
        timer = laserTime;
        laserPoints = getLaserPoints(6);
        laserPoints2 = getLaserPoints(6);
        damageFrame = true;
        previousTarget = target;
        previousTargetType = targetType;
    }
}

// def draw(self, surface, frame, color, hitboxes=False, offset_x=0, offset_y=0): ...
void Laser::draw(SDL_Renderer* renderer, const Frame& frame, Color color, bool hitboxesMode) {
    if (hitboxesMode) {
        for (auto& p : hitboxes) {
            Vec2 pos = frame.worldToScreen(p.x, p.y);
            int cx = static_cast<int>(pos.x);
            int cy = static_cast<int>(pos.y);
            int radius = std::max(2, static_cast<int>(frame.zoom * 2.0));
            Canvas::circle(renderer, { static_cast<double>(cx), static_cast<double>(cy) },
                            static_cast<double>(radius), color);
        }
        return;
    }

    for (auto* laserPartPtr : { &laserPoints, &laserPoints2 }) {
        auto& laserPart = *laserPartPtr;
        if (laserPart.empty()) continue; // defensive -- laser_points always has entries once tick() has run at least once
        double oglength = laserPart[laserPart.size() / 2];
        double scale = length / oglength;
        int halfSize = static_cast<int>(laserPart.size()) / 2;

        std::vector<Vec2> polygonPoints;
        polygonPoints.reserve(laserPart.size());
        for (size_t idx = 0; idx < laserPart.size(); ++idx) {
            double point = laserPart[idx];
            // was: if True or point <= self.length: -- always true, see
            // Laser.h class note; the else branch never executes.
            double waveHeight = thickness * std::sin((point + sinWaveOffset) * 1.5) * (0.5 + timer / laserTime);
            double x, y;
            if (static_cast<int>(idx) % halfSize == 0) {
                x = point * std::cos(angle) * scale;
                y = point * std::sin(angle) * scale;
            } else {
                x = point * std::cos(angle) * scale + waveHeight * std::sin(angle);
                y = point * std::sin(angle) * scale - waveHeight * std::cos(angle);
            }
            Vec2 screenPos = frame.worldToScreen(x + startX, y + startY);
            polygonPoints.push_back(screenPos);
        }

        if (polygonPoints.size() >= 3) {
            Canvas::polygon(renderer, polygonPoints, color);
        }
    }
}
